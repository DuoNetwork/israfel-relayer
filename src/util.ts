import { BigNumber } from '0x.js';
import moment from 'moment';
import * as os from 'os';
import * as CST from './constants';
import { IOption } from './types';

class Util {
	public logLevel: string = CST.LOG_INFO;

	public logInfo(text: any): void {
		this.log(text, CST.LOG_INFO);
	}

	public logDebug(text: any): void {
		this.log(text, CST.LOG_DEBUG);
	}

	public logError(text: any): void {
		this.log(text, CST.LOG_ERROR);
	}

	private log(text: any, level: string): void {
		if (CST.LOG_RANKING[this.logLevel] >= CST.LOG_RANKING[level])
			console.log(`${moment().format('HH:mm:ss.SSS')} [${level}]: ` + text);
	}

	public isNumber(input: any): boolean {
		const num = Number(input);
		return isFinite(num) && !isNaN(num);
	}

	public isEmptyObject(obj: object | undefined | null): boolean {
		if (!obj) return true;

		for (const prop in obj) if (obj.hasOwnProperty(prop)) return false;

		return true;
	}

	public getHostName() {
		return os.hostname();
	}

	public getRandomFutureDateInSeconds() {
		return new BigNumber(Date.now() + CST.TEN_MINUTES_MS).div(CST.ONE_SECOND_MS).ceil();
	}

	public stringToBN(value: string): BigNumber {
		return new BigNumber(value);
	}

	public defaultOption: IOption = {
		live: false,
		token: 'ZRX',
		amount: 1,
		maker: 0,
		spender: 1
	};

	public getUTCNowTimestamp() {
		return moment().valueOf();
	}

	public parseOptions(argv: string[]): IOption {
		const option: IOption = this.defaultOption;

		for (let i = 3; i < argv.length; i++) {
			option.live = process.argv.includes('live');
			const args = argv[i].split('=');
			switch (args[0]) {
				case 'token':
					option.token = args[1] || option.token;
					break;
				case 'amount':
					option.amount = Number(args[1]) || option.amount;
					break;
				case 'maker':
					option.maker = Number(args[1]) || option.maker;
					break;
				case 'spender':
					option.maker = Number(args[1]) || option.spender;
					break;
				default:
					break;
			}
		}

		return option;
	}
}

const util = new Util();
export default util;
