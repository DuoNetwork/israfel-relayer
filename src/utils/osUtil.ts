import * as os from 'os';

class OsUtil {
	public getHostName() {
		return os.hostname();
	}
}

const osUtil = new OsUtil();
export default osUtil;
