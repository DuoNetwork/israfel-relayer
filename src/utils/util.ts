import { Constants } from '@finbook/israfel-common';
import { IOption } from '../common/types';

export default class Util {
	public static defaultOption: IOption = {
		env: Constants.DB_DEV,
		tokens: [],
		token: '',
		debug: false,
		server: false
	};

	public static parseOptions(argv: string[]): IOption {
		const option: IOption = this.defaultOption;
		option.server = argv.includes('server');
		option.debug = argv.includes('debug');
		for (let i = 3; i < argv.length; i++) {
			const args = argv[i].split('=');
			switch (args[0]) {
				case 'env':
					option.env = [Constants.DB_LIVE, Constants.DB_UAT].includes(args[1])
						? args[1]
						: option.env;
					break;
				case 'tokens':
					option.tokens = args[1].split(',');
					break;
				case 'token':
					option.token = args[1] || option.token;
					break;
				default:
					break;
			}
		}

		return option;
	}
}
