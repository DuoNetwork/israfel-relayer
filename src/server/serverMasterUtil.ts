import child_process from 'child_process';
import { IOption, ISubProcess, IToken } from '../common/types';
import dynamoUtil from '../utils/dynamoUtil';
import util from '../utils/util';
import Web3Util from '../utils/Web3Util';

class ServerMasterUtil {
	public subProcesses: { [key: string]: ISubProcess } = {};

	public async startLaunching(
		web3Util: Web3Util,
		tool: string,
		option: IOption,
		startServer: (web3Util: Web3Util, option: IOption) => any
	): Promise<void> {
		const tokens: IToken[] = await dynamoUtil.scanTokens();

		if (!option.token) {
			util.logInfo('launching all pairs ' + tokens.map(token => token.code).join(','));
			for (const token of tokens) {
				await util.sleep(1000);
				this.subProcesses[token.code] = {
					token: token.code,
					lastFailTimestamp: 0,
					failCount: 0,
					instance: undefined as any
				};
				this.launchTokenPair(tool, token.code, option);
			}
		} else {
			const token = tokens.find(t => t.code === option.token);
			if (!token) throw new Error('invalid token specified');

			util.logInfo(
				`[${option.token}]:` + 'start launching orderBookServer for pair ' + option.token
			);
			startServer(web3Util, option);
		}
	}

	public launchTokenPair(tool: string, token: string, option: IOption) {
		const cmd =
			`npm run ${tool} token=${token} ${option.server ? ' server' : ''}${
				option.debug ? ' debug' : ''
			}` +
			(process.platform === 'win32' ? ' >>' : ' &>') +
			` ${tool}.${token}.log`;

		util.logInfo(`[${token}]: ${cmd}`);

		const procInstance = child_process.exec(
			cmd,
			process.platform === 'win32' ? {} : { shell: '/bin/bash' }
		);

		this.subProcesses[token].instance = procInstance;
		this.subProcesses[token].lastFailTimestamp = util.getUTCNowTimestamp();

		if (!procInstance) {
			util.logError('Failed to launch ' + token);
			this.retry(tool, option, token);
		} else {
			util.logInfo(`[${token}]: Launched ${tool}`);
			procInstance.on('exit', code => {
				util.logError(`[${token}]: Exit with code ${token}`);
				if (code) this.retry(tool, option, token);
			});
		}
	}

	public retry(tool: string, option: IOption, token: string) {
		const now: number = util.getUTCNowTimestamp();

		if (now - this.subProcesses[token].lastFailTimestamp < 30000)
			this.subProcesses[token].failCount++;
		else this.subProcesses[token].failCount = 1;

		this.subProcesses[token].lastFailTimestamp = now;

		if (this.subProcesses[token].failCount < 3)
			setTimeout(() => this.launchTokenPair(tool, token, option), 5000);
		else util.logError('Retry Aborted ' + token);
	}
}

const serverMasterUtil = new ServerMasterUtil();
export default serverMasterUtil;
