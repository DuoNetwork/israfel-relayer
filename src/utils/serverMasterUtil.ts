import child_process from 'child_process';
import { IOption, ISubProcess, IToken } from '../common/types';
import dynamoUtil from './dynamoUtil';
import osUtil from './osUtil';
import util from './util';

class ServerMasterUtil {
	public subProcesses: { [key: string]: ISubProcess } = {};

	public async startLaunching(
		tool: string,
		option: IOption,
		startServer?: (option: IOption) => any
	): Promise<void> {
		const tokens: IToken[] = await dynamoUtil.scanTokens();

		if (option.token && startServer) {
			const rawToken = tokens.find(t => t.code === option.token);
			if (!rawToken) throw new Error('invalid token specified');

			util.logInfo(`[${option.token}]: start launching ${tool}`);
			startServer(option);
		} else if (!option.tokens.length) {
			util.logInfo('launching all pairs ' + tokens.map(token => token.code).join(','));
			for (const token of tokens) {
				util.logInfo(`[${token.code}]: start launching ${tool}`);
				await util.sleep(1000);
				this.subProcesses[token.code] = {
					token: token.code,
					lastFailTimestamp: 0,
					failCount: 0,
					instance: undefined as any
				};
				this.launchTokenPair(tool, token.code, option);
			}
		} else if (option.tokens.length)
			for (const token of tokens) {
				if (!option.tokens.includes(token.code)) continue;

				util.logInfo(`[${token.code}]: start launching ${tool}`);
				await util.sleep(1000);
				this.subProcesses[token.code] = {
					token: token.code,
					lastFailTimestamp: 0,
					failCount: 0,
					instance: undefined as any
				};
				this.launchTokenPair(tool, token.code, option);
			}
	}

	public launchTokenPair(tool: string, token: string, option: IOption) {
		const cmd =
			`npm run ${tool} token=${token} env=${option.env} ${option.server ? ' server' : ''}${
				option.debug ? ' debug' : ''
			}` +
			(osUtil.isWindows() ? ' >>' : ' &>') +
			` ${tool}.${token}.log`;

		util.logInfo(`[${token}]: ${cmd}`);

		const procInstance = child_process.exec(
			cmd,
			osUtil.isWindows() ? {} : { shell: '/bin/bash' }
		);

		this.subProcesses[token].instance = procInstance;
		this.subProcesses[token].lastFailTimestamp = util.getUTCNowTimestamp();

		if (!procInstance) {
			util.logError('Failed to launch ' + token);
			this.retry(tool, option, token);
		} else {
			util.logInfo(`[${token}]: Launched ${tool}`);
			procInstance.on('exit', code => {
				util.logError(`[${token}]: Exit with code ${token}`);
				if (code) this.retry(tool, option, token);
			});
		}
	}

	public retry(tool: string, option: IOption, token: string) {
		const now: number = util.getUTCNowTimestamp();

		if (now - this.subProcesses[token].lastFailTimestamp < 30000)
			this.subProcesses[token].failCount++;
		else this.subProcesses[token].failCount = 1;

		this.subProcesses[token].lastFailTimestamp = now;

		if (this.subProcesses[token].failCount < 3)
			setTimeout(() => this.launchTokenPair(tool, token, option), 5000);
		else util.logError('Retry Aborted ' + token);
	}
}

const serverMasterUtil = new ServerMasterUtil();
export default serverMasterUtil;
