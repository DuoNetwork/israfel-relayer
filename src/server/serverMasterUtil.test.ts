import child_process from 'child_process';
import osUtil from '../utils/osUtil';
import util from '../utils/util';
import serverMasterUtil from './serverMasterUtil';

test('retry after long enought time', () => {
	child_process.exec = jest.fn() as any;
	util.getUTCNowTimestamp = jest.fn(() => 1234567890);
	serverMasterUtil.subProcesses['token'] = {
		token: 'token',
		lastFailTimestamp: util.getUTCNowTimestamp() - (30000 + 1),
		failCount: 2,
		instance: undefined as any
	};
	serverMasterUtil.retry(
		'tool',
		{
			server: false,
			debug: false,
			live: false
		} as any,
		'token'
	);
	expect(serverMasterUtil.subProcesses['token']).toMatchSnapshot();
});

test('retry within short time', () => {
	child_process.exec = jest.fn() as any;
	util.getUTCNowTimestamp = jest.fn(() => 1234567890);
	serverMasterUtil.subProcesses['token'] = {
		token: 'token',
		lastFailTimestamp: util.getUTCNowTimestamp() - (30000 - 1),
		failCount: 2,
		instance: undefined as any
	};
	serverMasterUtil.retry(
		'tool',
		{
			debug: false,
			live: false
		} as any,
		'token'
	);
	expect(serverMasterUtil.subProcesses['token']).toMatchSnapshot();
});

test('launchTokenPair fail win32', () => {
	osUtil.isWindows = jest.fn(() => true);
	child_process.exec = jest.fn() as any;
	util.getUTCNowTimestamp = jest.fn(() => 1234567890);
	serverMasterUtil.retry = jest.fn();
	serverMasterUtil.subProcesses['token'] = {
		token: 'token',
		lastFailTimestamp: 0,
		failCount: 0,
		instance: undefined as any
	};
	serverMasterUtil.launchTokenPair('tool', 'token', {
		debug: false,
		live: false
	} as any);
	expect(((child_process.exec as any) as jest.Mock<Promise<void>>).mock.calls).toMatchSnapshot();
	expect(serverMasterUtil.subProcesses).toMatchSnapshot();
	expect((serverMasterUtil.retry as jest.Mock<void>).mock.calls).toMatchSnapshot();
});

test('launchTokenPair success win 32', () => {
	osUtil.isWindows = jest.fn(() => true);
	child_process.exec = jest.fn(() => {
		return {
			on: jest.fn()
		};
	}) as any;
	util.getUTCNowTimestamp = jest.fn(() => 1234567890);
	serverMasterUtil.retry = jest.fn();
	serverMasterUtil.subProcesses['token'] = {
		token: 'token',
		lastFailTimestamp: 5,
		failCount: 0,
		instance: undefined as any
	};
	serverMasterUtil.launchTokenPair('tool', 'token',  {
		debug: false,
		live: false
	} as any);
	expect(((child_process.exec as any) as jest.Mock<Promise<void>>).mock.calls).toMatchSnapshot();
	expect(serverMasterUtil.subProcesses).toMatchSnapshot();
	expect(
		(serverMasterUtil.subProcesses['token'].instance.on as jest.Mock<void>).mock.calls
	).toMatchSnapshot();
});

test('launchTokenPair debug win32', () => {
	osUtil.isWindows = jest.fn(() => true);
	child_process.exec = jest.fn() as any;
	util.getUTCNowTimestamp = jest.fn(() => 1234567890);
	serverMasterUtil.subProcesses['token'] = {
		token: 'token',
		lastFailTimestamp: 0,
		failCount: 0,
		instance: undefined as any
	};
	serverMasterUtil.launchTokenPair('tool', 'token',  {
		debug: true,
		live: false
	} as any);
	expect(((child_process.exec as any) as jest.Mock<Promise<void>>).mock.calls).toMatchSnapshot();
	expect(serverMasterUtil.subProcesses).toMatchSnapshot();
});

test('launchTokenPair fail not win32', () => {
	osUtil.isWindows = jest.fn(() => false);
	child_process.exec = jest.fn() as any;
	util.getUTCNowTimestamp = jest.fn(() => 1234567890);
	serverMasterUtil.retry = jest.fn();
	serverMasterUtil.subProcesses['token'] = {
		token: 'token',
		lastFailTimestamp: 0,
		failCount: 0,
		instance: undefined as any
	};
	serverMasterUtil.launchTokenPair('tool', 'token', {
		debug: false,
		live: false
	} as any);
	expect(((child_process.exec as any) as jest.Mock<Promise<void>>).mock.calls).toMatchSnapshot();
	expect(serverMasterUtil.subProcesses).toMatchSnapshot();
	expect((serverMasterUtil.retry as jest.Mock<void>).mock.calls).toMatchSnapshot();
});

test('launchTokenPair success not win32', () => {
	osUtil.isWindows = jest.fn(() => false);
	child_process.exec = jest.fn(() => {
		return {
			on: jest.fn()
		};
	}) as any;
	util.getUTCNowTimestamp = jest.fn(() => 1234567890);
	serverMasterUtil.retry = jest.fn();
	serverMasterUtil.subProcesses['token'] = {
		token: 'token',
		lastFailTimestamp: 5,
		failCount: 0,
		instance: undefined as any
	};
	serverMasterUtil.launchTokenPair('tool', 'token',  {
		debug: false,
		live: false
	} as any);
	expect(((child_process.exec as any) as jest.Mock<Promise<void>>).mock.calls).toMatchSnapshot();
	expect(serverMasterUtil.subProcesses).toMatchSnapshot();
	expect(
		(serverMasterUtil.subProcesses['token'].instance.on as jest.Mock<void>).mock.calls).toMatchSnapshot()
});

test('launchTkenPair debug not win32', () => {
	osUtil.isWindows = jest.fn(() => false);
	child_process.exec = jest.fn() as any;
	util.getUTCNowTimestamp = jest.fn(() => 1234567890);
	serverMasterUtil.subProcesses['token'] = {
		token: 'token',
		lastFailTimestamp: 0,
		failCount: 0,
		instance: undefined as any
	};
	serverMasterUtil.launchTokenPair('tool', 'token',  {
		debug: true,
		live: false
	} as any);
	expect(((child_process.exec as any) as jest.Mock<Promise<void>>).mock.calls).toMatchSnapshot();
	expect(serverMasterUtil.subProcesses).toMatchSnapshot();
});
