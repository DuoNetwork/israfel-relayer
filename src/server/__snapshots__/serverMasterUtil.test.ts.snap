// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`launchTokenPair debug 1`] = `
Array [
  Array [
    "npm run tool token=token  debug >> tool.token.log",
    Object {},
  ],
]
`;

exports[`launchTokenPair debug 2`] = `
Object {
  "token": Object {
    "failCount": 0,
    "instance": undefined,
    "lastFailTimestamp": 1234567890,
    "token": "token",
  },
}
`;

exports[`launchTokenPair fail 1`] = `
Array [
  Array [
    "npm run tool token=token  >> tool.token.log",
    Object {},
  ],
]
`;

exports[`launchTokenPair fail 2`] = `
Object {
  "token": Object {
    "failCount": 0,
    "instance": undefined,
    "lastFailTimestamp": 1234567890,
    "token": "token",
  },
}
`;

exports[`launchTokenPair fail 3`] = `
Array [
  Array [
    "tool",
    Object {
      "debug": false,
      "live": false,
    },
    "token",
  ],
]
`;

exports[`launchTokenPair success 1`] = `
Array [
  Array [
    "npm run tool token=token  >> tool.token.log",
    Object {},
  ],
]
`;

exports[`launchTokenPair success 2`] = `
Object {
  "token": Object {
    "failCount": 0,
    "instance": Object {
      "on": [MockFunction] {
        "calls": Array [
          Array [
            "exit",
            [Function],
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      },
    },
    "lastFailTimestamp": 1234567890,
    "token": "token",
  },
}
`;

exports[`launchTokenPair success 3`] = `
Array [
  Array [
    "exit",
    [Function],
  ],
]
`;

exports[`retry after long enought time 1`] = `
Object {
  "failCount": 1,
  "instance": undefined,
  "lastFailTimestamp": 1234567890,
  "token": "token",
}
`;

exports[`retry within short time 1`] = `
Object {
  "failCount": 3,
  "instance": undefined,
  "lastFailTimestamp": 1234567890,
  "token": "token",
}
`;
